sstables/sstables.cc

```diff
@@ -2679,20 +2679,24 @@ future<> sstable::set_generation(int64_t new_generation) {
     });
 }
 
-void sstable::move_to_new_dir_in_thread(sstring new_dir, int64_t new_generation) {
-    create_links(new_dir, new_generation).get();
-    remove_file(filename(component_type::TOC)).get();
-    sstable_write_io_check(sync_directory, _dir).get();
-    sstring old_dir = std::exchange(_dir, std::move(new_dir));
-    int64_t old_generation = std::exchange(_generation, new_generation);
-    parallel_for_each(all_components(), [this, old_generation, old_dir] (auto p) {
-        if (p.first == component_type::TOC) {
-            return make_ready_future<>();
-        }
-        return remove_file(sstable::filename(old_dir, _schema->ks_name(), _schema->cf_name(), _version, old_generation, _format, p.second));
-    }).get();
-    sync_directory(_dir).get();
-    sync_directory(old_dir).get();
+future<> sstable::move_to_new_dir(sstring new_dir, int64_t new_generation) {
+    sstring old_dir = get_dir();
+    return create_links(new_dir, new_generation).then([this] {
+        return remove_file(filename(component_type::TOC));
+    }).then([this] {
+        return sstable_write_io_check(sync_directory, _dir);
+    }).then([this, old_dir, new_dir, new_generation] {
+        _dir = new_dir;
+        int64_t old_generation = std::exchange(_generation, new_generation);
+        return parallel_for_each(all_components(), [this, old_generation, old_dir] (auto p) {
+            if (p.first == component_type::TOC) {
+                return make_ready_future<>();
+            }
+            return remove_file(sstable::filename(old_dir, _schema->ks_name(), _schema->cf_name(), _version, old_generation, _format, p.second));
+        });
+    }).then([this, old_dir, new_dir] {
```

Comment:
[no action needed] by now you could std::move the string, but perhaps it's better to leave it as is, in case somebody adds a continuation later.